multidim-interop/versions.ts

```diff
@@ -1,195 +1,193 @@
-import gov028 from "./impl/go/v0.28/image.json"
-import gov027 from "./impl/go/v0.27/image.json"
-import gov026 from "./impl/go/v0.26/image.json"
-import gov025 from "./impl/go/v0.25/image.json"
-import gov024 from "./impl/go/v0.24/image.json"
-import gov023 from "./impl/go/v0.23/image.json"
-import gov022 from "./impl/go/v0.22/image.json"
-import rustv048 from "./impl/rust/v0.48/image.json"
-import rustv049 from "./impl/rust/v0.49/image.json"
-import rustv050 from "./impl/rust/v0.50/image.json"
-import rustv051 from "./impl/rust/v0.51/image.json"
-import jsV041 from "./impl/js/v0.41/node-image.json"
-import jsV042 from "./impl/js/v0.42/node-image.json"
-import jsV044 from "./impl/js/v0.44/node-image.json"
-import jsV045 from "./impl/js/v0.45/image.json"
-import nimv10 from "./impl/nim/v1.0/image.json"
-import chromiumJsV041 from "./impl/js/v0.41/chromium-image.json"
-import chromiumJsV042 from "./impl/js/v0.42/chromium-image.json"
-import chromiumJsV044 from "./impl/js/v0.44/chromium-image.json"
-import chromiumJsV045 from "./impl/js/v0.45/chromium-image.json"
-import firefoxJsV045 from "./impl/js/v0.45/firefox-image.json"
-import zigv001 from "./impl/zig/v0.0.1/image.json"
-import javav001 from "./impl/java/v0.0.1/image.json"
+import fs from "fs"
 
 export type Version = {
     id: string,
-    containerImageID: string,
+    // This can be the image ID, or a function that takes the version ID and returns the image ID
+    containerImageID: string | ((id: string) => string),
     // If defined, this will increase the timeout for tests using this version
     timeoutSecs?: number,
     transports: Array<(string | { name: string, onlyDial: boolean })>,
     secureChannels: string[],
     muxers: string[]
 }
 
+// Loads the container image id for the given version id. Expects the form of
+// "<impl>-vX.Y.Z" or "<impl>vX.Y" and the image id to be in the file
+// "./impl/<impl>/vX.Y/image.json" or "./impl/<impl>/v0.0.Z/image.json"
+function canonicalImageIDLookup(id: string): string {
+    // Split by implementation and version
+    const [impl, version] = id.split("-v")
+    // Drop the patch version
+    const [major, minor, patch] = version.split(".")
+    let versionFolder = `v${major}.${minor}`
+    if (major === "0" && minor === "0") {
+        // We're still in the 0.0.x phase, so we use the patch version
+        versionFolder = `v0.0.${patch}`
+    }
+    // Read the image ID from the JSON file on the filesystem
+    const imageIDJSON = fs.readFileSync(`./impl/${impl}/${versionFolder}/image.json`, "utf8")
+    const imageID = JSON.parse(imageIDJSON).imageID
+    return imageID
+}
+
 export const versions: Array<Version> = [
     {
         id: "rust-v0.48.0",
-        containerImageID: rustv048.imageID,
+        containerImageID: canonicalImageIDLookup,
```

Comment:
Can this be made the default, i.e. use `canonicalImageIDLookup` if unset, so you don't have to define it here 50 times?

Reply:
Good idea!

Reply:
I thought about making it default, but I felt like it obfuscated what was happening a bit more than being explicit. Think of someone looking at this file for the first time. The containerImageID can be a string or it can be a function that returns a string. If it's a function you can go directly to that function at look at what it does. Straightforward, and nothing hidden. If we make this default, there's a missing piece between where the version is defined and how the containerImageID is defined.

If I'm in the minority opinion here, I'm fine to make this change. Just want to outline my thoughts.